cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TypeManager.kt

```diff
@@ -187,13 +190,35 @@ class TypeManager {
         return parameterizedType
     }
 
-    fun <T : Type> registerType(t: T): T {
+    inline fun <reified T : Type> registerType(t: T): T {
+        // Skip as they should be unique to each class and not globally unique
+        if (t is ParameterizedType) {
+            return t
+        }
+
         if (t.isFirstOrderType) {
-            firstOrderTypes.add(t)
-        } else {
-            secondOrderTypes.add(t)
-            registerType((t as SecondOrderType).elementType)
+            // Make sure we only ever return one unique object per type
+            if (!firstOrderTypes.add(t)) {
```

Comment:
Does this not add the element if already contained? I can't find any Doc on KeySets saying that it functions like this.

Reply:
The doc says:
```
/**
     * Adds the specified element to the set.
     *
     * @return `true` if the element has been added, `false` if the element is already contained in the set.
     */
```

And since it is a `Set`, it will only contain each "equal" object exactly once.